odoo/models.py

```diff
@@ -4506,6 +4518,82 @@ def _parent_store_update(self):
+            having matching data in search_fields
+
+            :param data_list: list of dicts with keys `xml_id` and record values
+            :param search_fields: list of technical field names
+
+            :return: updated data_list containing the new/found xml_id
+
+            :raises ValidationError: when multiple records match the search criteria
+        '''
+        # check if search_fields exist in the models and filter out everything else preventing error and SQL injection
+        search_fields = [f for f in search_fields if f in self._fields.keys()]
+        if not search_fields:
+            return data_list
+
+        key_fields = ', '.join(search_fields)
+        prefixed_key_fields = ', '.join([f'r.{field}' for field in search_fields])
+        key_data = [tuple(item['values'][field] for field in search_fields) for item in data_list]
+
+        sql = f"""
+            CREATE TEMP TABLE temp_{self._table}(
+                {', '.join([f'{f} {self._fields[f].column_type[1]}' for f in search_fields])}
+            );
+            INSERT INTO temp_{self._table}
+            VALUES {','.join(['%s' for d in key_data])};
+            WITH rec_count AS (
+                SELECT
+                    ({key_fields}) AS val_tup,
+                    count(*) as cnt
+                FROM
+                    {self._table}
+                {f'WHERE company_id IS NULL or company_id = {self.env.company.id} ' if 'company_id' in self._fields else None}
+                GROUP BY val_tup
+            ), tmp_data AS (
+                SELECT
+                    DISTINCT ({key_fields}) AS val_tup
+                FROM temp_{self._table}
+            )
+            SELECT
+                {prefixed_key_fields},
+                r.id,
+                i.module || '.' || i.name as external_id,
+                c.cnt
+            FROM {self._table} r
+            INNER JOIN tmp_data t ON ({prefixed_key_fields}) = t.val_tup
+            LEFT JOIN rec_count c ON ({prefixed_key_fields}) = c.val_tup
+            LEFT JOIN ir_model_data i ON r.id = i.res_id AND i.model = '{self._name}'
+            {f'WHERE r.company_id IS NULL or company_id = {self.env.company.id} ' if 'company_id' in self._fields else None}
+        """
+        self.env.cr.execute(sql, tuple(key_data))
+        res = self.env.cr.fetchall()
+        self.env.cr.execute(f'DROP TABLE temp_{self._table}')
+        res_dict = {
+            row[:len(search_fields)]: {
+                'id': row[-3],
+                'xid': row[-2],
+                'count': row[-1],
+            } for row in res if row[-1] == 1
+        }
+        if len(res_dict) < len(res):
+            raise ValidationError(_('Multiple records found matching search key criteria'))
+        records_without_xid = [row['id'] for row in res_dict.values() if row['id'] and not row['xid']]
+        new_external_ids = {}
+        for rec, xid in self.browse(records_without_xid).__ensure_xml_id():
```

Comment:
why do you need to create additional XML IDs? The whole point of the feature is to avoid using XML IDs at all, since the import keys replace it

Reply:
I thought it was required - I guess I should just read more :)